### PR TITLE
fix(ledger,lfs): self-heal stale git locks, close two silent recording-loss paths

### PR DESCRIPTION
**Pull request by rsnodgrass**
## Summary

Ledger repos can wedge permanently. This PR ships the two repairs that are unambiguously safe, and documents the root cause of the larger wedge without attempting it yet — an adversarial review found that the obvious fix would trade a loud, harmless wedge for silent data loss.

**Found while investigating five stuck ledgers on one machine:**

| Ledger | State |
| --- | --- |
| `repo_019c6d2e` | **341 ahead / 1055 behind**, 13 days, 281 conflicts |
| `repo_019c7d5c` | unborn HEAD, 184 uncommitted files, doctor reports **skipped** |
| `repo_019ddf8e` | `index.lock` + `next-index-13088.lock` from **May 3** (~3 months) |
| `repo_019f4537`, `repo_019ce879` | diverged 39/35 and 2/1 |

## What this PR ships

- **Stale git locks now self-heal on the push path.** `IsSafeForGitOps` only ever *reported* locks; it never cleared them — while the pull path has swept them before every pull for ages. That asymmetry means a lock left by a crashed git blocks every future push forever. `IsSafeForGitOps` now sweeps first, still gated on the existing 5-minute `StaleLockAge` so a live git process is never disturbed.
- **`next-index-<pid>.lock` is now covered.** Its pid varies, so it could never be listed in `knownLockFiles` — it was undetectable *and* unclearable. Detection and removal now share one helper (`lockFilesIn`) so they can't drift again. Deliberately not a broad `*.lock`, which would eat `packed-refs.lock` and ref-transaction locks.
- **`.needs-summary` stops entering git.** `sessions/.gitignore` was three comment lines that ignored nothing, so the marker got swept in — 111 of them tracked on one real ledger. Its payload is *absolute local paths* (`{"cache_dir":"/Users/<name>/…"}`), so it leaks the writer's home directory into a shared repo and can never merge cleanly between two machines. Already-tracked copies are untracked via `git rm --cached` alongside the existing `*.rej` handling; local files are preserved, so in-flight summarization is unaffected.

Together these unstick the lock-blocked ledger outright and remove 12 of the 281 conflicts on the big one.

## The wedge itself, now fixed

`sessions/<id>/meta.json` is written by **both** the cloud summarizer and the local CLI. Every cycle:

```mermaid
flowchart LR
    A["git pull rebase"] --> B["halts on<br/>sessions/*/meta.json"]
    B --> C["accept-theirs refuses:<br/>not a safe prefix"]
    C --> D["LLM tier:<br/>ErrLLMUnavailable"]
    D --> E["AuditAndAbort<br/>rolls back"]
    E --> F["state unchanged,<br/>still diverged"]
    F --> A
```

Zero state change, zero escalation, forever.

- **`sessions/` now auto-resolves** — added to `ledger.DefaultResolveRules`, which every push/pull site already funnels through. Session artifacts are best-effort or regenerable, and the forward links are documented caches whose canonical source is elsewhere. An imperfect summary beats a ledger that can never sync.
- **Guarded by pointer-wins.** During a rebase, stage 2 is the cloud and stage 3 is the local commit being replayed, so `checkout --theirs` picks **local** — the opposite of what the name suggests (verified in a scratch repo, not inferred). On the wedged ledger every mixed content conflict is `local=hydrated / remote=pointer`, so the naive fix would commit hydrated bytes over the pointer and permanently break pushes. Now: when two sides differ in *kind*, the LFS pointer wins. That rule is commutative and idempotent, so unlike the positional rule it **converges** rather than ping-ponging between replicas.
- **Found by the new harness — a real add/add bug.** When both sides *create* the same path (no merge base), the conflict skipped the content branch entirely and got staged **with `<<<<<<< HEAD` markers committed into the file**. That is the common shape when both writers produce a new `session.md` for the same session. Now handled by the same pointer-wins path.
- **`fixLedgerBranchBehind` routes through `automerge`** — it was the one reconcile path bypassing the tiered resolver.

Also worth knowing: `resolveHardDenyList` is **inert for ledgers** — its only consumer is `ValidateResolveRules` via `manifest/parser.go`, and no ledger has a `sync.manifest`. Editing it would have changed nothing.

## Two P0 data-loss paths, now closed

Both were **pre-existing** and independent of the wedge — but unwedging `sessions/` would have unlocked the second, turning a loud harmless wedge into a silent shredder. Fixing them is a prerequisite, so they are in this PR.

- **`WritePointerFile` overwrote blobs with no OID compare and no backup.** `UpdateMetaSummary` rewrites pointers for the *whole* `Files` map, so a single stale OID — exactly what independently-resolved merge conflicts produce — destroyed the only local copy of a recording. If that OID was never uploaded, the push was then rejected and the reconcile path blanked the file to zero bytes. **No doctor check detects the result** (`session_manifest.go` gates its size-0 and pointer-shape checks on `ref.IsGit()`, so a hydrated or 0-byte LFS entry passes clean). It now refuses unless the on-disk bytes hash to the OID — which is what the real upload path always produces (`meta.go:818`), so the happy path is untouched.
- **`ReconcileUnpushedPointers` treated *any* LFS batch error as "blob is gone"** and wrote `[]byte{}`. A transient 401, 429, or 5xx blanked live recordings. Only 404 counts as missing now; anything else aborts the reconcile. A push that stays blocked is recoverable; `[]byte{}` is the one operation with no inverse.

Several tests asserted the old destructive behavior using synthetic OIDs (`sha256:abc123`) over content that hashes to something else — a physically impossible state. They now use real content hashes, which is both faithful to production and what makes the guard meaningful.

## Still deferred (`ox-zmjc` epic)

The `sessions/` auto-resolve itself, which needs a **pointer-wins guard**: when two sides of a conflict differ in *kind* (pointer vs. bytes), take the pointer. Unlike positional accept-theirs, that rule is commutative and idempotent, so it actually converges. Also filed: the `session_id` split-brain, where 26 of 33 conflicts differ *only* in that field.

## Test Plan

**New harness: `internal/gitutil/wedge_harness_test.go`.** A reusable real-git fixture (bare remote + local clone + a second clone standing in for the cloud), with one case per wedge class that reached production. Every case asserts the same three-part contract, and the third part is the one that matters:

1. the wedge is **detected**
2. the repair **clears** it
3. the ledger **makes progress** afterward — the next pull+push actually lands

Part 3 is why these bugs shipped: `git rebase --abort` returning success while the ledger stays permanently un-syncable is exactly the 341-ahead/1055-behind wedge. Cases cover: `sessions/` meta conflict self-heal, repair idempotence (no infinite ping-pong), pointer-never-loses-to-hydrated from *either* side, stale locks (`index.lock`, `next-index-<pid>.lock`, `shallow.lock`), stale rebase dir, fresh-rebase-is-never-aborted, divergence convergence to 0/0, bounded rounds under 25 simultaneous conflicts, and human-authored paths still refusing auto-resolve.

- `internal/daemon` — new `TestSessionMetaConflict_AutoResolvesUnderProductionRules` fails if `sessions/` ever drops out of the production rules. The existing restart-escalation test was pinned to the old "sessions/ stays hard-denied" assumption; it now drives non-resolving rules so it still tests the escalation math it was written for.
- `internal/gitutil` — pointer-wins classification, LFS-pointer detection edge cases (oversized-with-prefix, spec URL mentioned mid-body), lock sweep.
- `internal/lfs` — pointer-overwrite refusal, redaction-style pointer replacement still allowed, sessions `.gitignore` content + idempotence.
- 102 packages green, `make lint` 0 issues.
- Verified against a **copy** of the real wedged ledger: 348 ahead / 1062 behind, 47 merge-tree conflicts, 173 pre-existing hydrated/pointer pairs. (The original was never mutated.)

---
<details>
<summary>Checklist (expand if needed)</summary>

- [x] Tests added
- [ ] CHANGELOG entry — N/A, internal repair paths, no user-facing surface change
- [x] Breaking change documented — none
- [x] `/security-review` — N/A for this diff (no `internal/auth/`, `internal/mcp/`, `raw_writer.go`, `prepush_scan.go`, `internal/upgrade/`, or `go.sum` changes). Note the two P0 LFS data-loss follow-ups filed above are pre-existing and want a look.

</details>

Co-Authored-By: [SageOx](https://github.com/SageOx)

### CODE REVIEW

File: internal/gitutil/safety.go [hunk 1]
```diff
@@ -130,6 +160,15 @@ func RebaseAge(repoPath string) (time.Duration, bool) {
...
 	gitDir := filepath.Join(repoPath, ".git")
 
+	// Sweep abandoned locks before reporting them as blocking. Without this the
+	// push path is asymmetric with the pull path (which already sweeps before
+	// every pull), so a lock left by a crashed git blocks every future push
+	// forever with no self-heal. Gated on StaleLockAge, so a lock held by a
+	// live git process is never touched.
+	if removed, errs := RemoveStaleLockFiles(gitDir); len(removed) > 0 || len(errs) > 0 {
+		slog.Info("git lock sweep", "repo", repoPath, "removed", strings.Join(removed, ","), "errors", len(errs))
+	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/gitutil/safety.go:251]: Fixed in eb28c806.

Valid finding — age alone was the only gate, and you are right that it can outlive a live owner.

**Owner liveness now takes precedence over age where an owner is knowable.** `next-index-<pid>.lock` encodes the PID that created it, so `RemoveStaleLockFiles` now resolves that PID and probes it (`os.FindProcess` + signal 0, with `ErrPermission` treated as alive since that means running-but-owned-by-another-user). A live owner retains the lock **regardless of age**, and `IsSafeForGitOps` keeps blocking. The probe is deliberately biased toward "alive": a false positive just leaves the lock for the next sweep, while a false negative is the index-corruption case you describe.

One correction on scope: `index.lock`, `shallow.lock`, `config.lock`, and `HEAD.lock` carry **no owner information at all** — git's `index.lock` *is* the lock, and its content is the partially-written index rather than metadata. There is no PID to validate, so age remains the only signal git's on-disk format offers for those. That is now stated explicitly at `lockOwnerPID` so the asymmetry does not read as an oversight. The 5-minute `StaleLockAge` and its use on the pull path both predate this PR; the change here made the push path consistent with the pull path, which had been sweeping before every pull already.

**Tests** (`TestLockSweep_OwnerLivenessBeatsAge`) now cover exactly the split you asked for:
- live owner (our own PID) + 100x `StaleLockAge` → **retained**, and `IsSafeForGitOps` still returns an error
- verifiably-dead owner → removed
- ownerless lock past the threshold → removed via the age fallback
- ownerless lock under the threshold → retained

Plus `TestLockOwnerPID` (including malformed/zero/negative PIDs) and `TestProcessAlive`.

File: internal/gitutil/safety.go [hunk 2]
```diff
@@ -130,6 +218,15 @@ func RebaseAge(repoPath string) (time.Duration, bool) {
 func IsSafeForGitOps(repoPath string) error {
 	gitDir := filepath.Join(repoPath, ".git")
 
+	// Sweep abandoned locks before reporting them as blocking. Without this the
+	// push path is asymmetric with the pull path (which already sweeps before
+	// every pull), so a lock left by a crashed git blocks every future push
+	// forever with no self-heal. Gated on StaleLockAge, so a lock held by a
+	// live git process is never touched.
+	if removed, errs := RemoveStaleLockFiles(gitDir); len(removed) > 0 || len(errs) > 0 {
+		slog.Info("git lock sweep", "repo", repoPath, "removed", strings.Join(removed, ","), "errors", len(errs))
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/gitutil/safety.go:250]: Fixed in e4ce49d3. Your repro is right and my earlier reply under-solved this.

I had argued that `index.lock` carries no owner so age is all we have. True, but I left the threshold at `StaleLockAge` (5 min) — and as you show, a slow but legitimate `git pull --rebase` on a large repo over a bad network really can hold the index longer than that. Sweeping there deletes an active lock and admits a second writer.

**Ownerless locks are now gated on a separate, much longer `AbandonedLockAge` (1 hour).** No live index operation reaches an hour; the real incident was an `index.lock` plus a `next-index-<pid>.lock` untouched for three months, so recovery still works while a live writer is never displaced. The split is now explicit:

- owner knowable (`next-index-<pid>.lock`) → **liveness decides, overriding age entirely**
- owner verifiably gone → removable after `StaleLockAge`
- ownerless (`index.lock`, `shallow.lock`, `config.lock`, `HEAD.lock`) → removable only after `AbandonedLockAge`

New `TestLockSweep_OwnerlessLockSurvivesTheStaleWindow` covers exactly your scenario for all four ownerless locks: held at 3x `StaleLockAge`, the sweep must remove nothing, the file must survive, and `IsSafeForGitOps` must still return an error rather than letting the push proceed. The pre-existing tests that asserted removal at the 5-minute mark were updated to the new contract rather than left pinning the unsafe behavior.

File: internal/gitutil/rebase_resolve.go [hunk 3]
```diff
@@ -133,6 +171,68 @@ func ResolveRebaseAcceptTheirs(ctx context.Context, repoPath string, safePrefixe
...
+const lfsPointerPrefix = "version https://git-lfs.github.com/spec/v1"
+
+// maxLFSPointerSize bounds how many bytes we read to classify a blob. Real
+// pointers are ~130 bytes; anything larger is content by definition.
+const maxLFSPointerSize = 200
+
+// isLFSPointerBlob reports whether the given blob content is an LFS pointer.
+func isLFSPointerBlob(content string) bool {
+	return len(content) <= maxLFSPointerSize && strings.HasPrefix(content, lfsPointerPrefix)
+}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/gitutil/rebase_resolve.go:223]: Fixed in e4ce49d3. Valid — the classifier was prefix + length only, so a truncated pointer would have won against valid content and then failed downstream LFS parsing. That is data loss wearing a safety guard.

`isLFSPointerBlob` now requires the complete pointer shape before pointer-wins applies: an `oid` line with a `sha256:` prefix and a non-empty, whitespace-free digest, plus a `size` line that parses as a positive integer. Anything malformed falls through to the positional rule, which is the correct conservative outcome — it treats the blob as ordinary content rather than granting it an automatic win.

This mirrors `lfs.ParsePointer`; it is duplicated rather than imported because `internal/lfs` depends on `gitutil`, so importing back would cycle. That constraint is noted at the constant.

`TestIsLFSPointerBlob` gained cases for: version line only, missing `oid`, missing `size`, empty digest, `oid` without the algorithm prefix, non-numeric size, zero size, and negative size — plus the pre-existing oversized-with-valid-prefix and spec-URL-mentioned-mid-body cases.
